### PR TITLE
Implement more linux/amd64 syscalls

### PR DESCRIPTION
**Pull request by samhocevar**

### CODE REVIEW

**Comment by hajimehoshi** [comment]
Could you tell me more context about this PR?

File: 1.19_linux/runtime/sys_linux_amd64.s.patch [hunk 1]
```diff
@@ -347,65 +347,65 @@ TEXT runtime·sysUnusedOS_trampoline(SB),NOSPLIT,$0
...
 TEXT runtime·sysMapOS_trampoline(SB),NOSPLIT,$0
 	PUSHQ	BP
 	MOVQ	SP, BP
 	MOVQ	8(DI), SI
 	MOVQ	0(DI), DI
-        CALL	hitsumabushi_sysMapOS(SB)
+	CALL	hitsumabushi_sysMapOS(SB)
 	POPQ	BP
-        RET
+	RET
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/sys_linux_amd64.s.patch]: Optional: indentation fix can be in another PR

**Comment by hajimehoshi** [comment]
Friendly ping

**Comment by samhocevar** [comment]
Hi! I’m sorry I’m still a bit struggling with a nasty Covid, but I’ll try to be clear.

This PR (with a few additional minor fixes) allows building the Go runtime with the native PS4 and PS5 compilers (also the Switch compiler). It also moves the special logic for file descriptors 0/1/2 to a single place.

The idea is then to let the user reimplement `c_write()`, `c_fnctl()` etc. with their implementation of choice.

Right now I can run a simple Lua-based application on the PlaysStation.

I understand that makes a lot of syscalls handled in `syscall_linux.go`. Maybe you’d prefer this to remain a very simple wrapper and have it implemented as a couple of large syscall/syscall6 C functions that the user can override?

File: 1.19_linux/runtime/sys_linux_arm64.s.patch [hunk 2]
```diff
@@ -503,3 +503,49 @@ TEXT runtime·sysMapOS_trampoline(SB),NOSPLIT,$0
...
+	MOVD	R0, R19		// R19 is callee-save
+	MOVD	8(R0), R1
+	MOVD	16(R0), R2
+	MOVW	24(R0), R3
+	MOVD	0(R0), R0
+	BL	c_renameat(SB)
+	MOVW	R0, 32(R19)	// return value
+	RET
+
+TEXT runtime·fstatat_trampoline(SB),NOSPLIT,$0
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/sys_linux_arm64.s.patch:543]: Let's order the functions in a consistent way with _amd64 version.

File: 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c [hunk 3]
```diff
@@ -265,12 +267,47 @@ int32_t c_write1(uintptr_t fd, void *p, int32_t n) {
     break;
   default:
     fprintf(stderr, "syscall write(%lu, %p, %d) is not implemented\n", fd, p, n);
+    ret = -ENOSYS;
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c]: EBADF?

https://man7.org/linux/man-pages/man2/write.2.html

File: 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c [hunk 4]
```diff
@@ -265,12 +267,47 @@ int32_t c_write1(uintptr_t fd, void *p, int32_t n) {
...
+
+int c_fcntl(int fd, int cmd, int arg)
+{
+  if (fd == 0 || fd == 1 || fd == 2) {
+    if (cmd == F_GETFL) {
+      return 0;
+    }
+  }
+  fprintf(stderr, "syscall fcntl(%d, %d, %d) is not implemented\n", fd, cmd, arg);
+  return -ENOSYS;
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c]: EBADF?

**Comment by hajimehoshi** [comment]
> The idea is then to let the user reimplement c_write(), c_fnctl() etc. with their implementation of choice.

I see. So we need more changes to add an API to provide the filename to overwrite later, right?

**Comment by samhocevar** [comment]
> > The idea is then to let the user reimplement c_write(), c_fnctl() etc. with their implementation of choice.
> 
> I see. So we need more changes to add an API to provide the filename to overwrite later, right?

Yes, that would be the next step. I just don’t know about the granularity, because maybe not everything should go to the same place. All filesystem functions could go to the same file, but for instance `c_sched_getaffinity()` is maybe more related to `c_futex()`, but I’m not sure about `c_walltime1()`.

File: 1.19_linux/runtime/sys_linux_arm64.s.patch (hunk 2 above)
Comment (samhocevar) [reply 1.19_linux/runtime/sys_linux_arm64.s.patch:543]: Done!

File: 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c (hunk 3 above)
Comment (samhocevar) [reply 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c]: Yes, no strong opinion but it makes sense.

File: 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c (hunk 4 above)
Comment (samhocevar) [reply 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c]: Done!

**Comment by samhocevar** [comment]
Should the functions be named `hitsumabushi_open()` etc. instead of `c_open()` etc.?

**Comment by hajimehoshi** [comment]
> Should the functions be named hitsumabushi_open() etc. instead of c_open() etc.?

Yes I think so.

**Comment by samhocevar** [comment]
So, this is what I would like to suggest:
 - move the following functions to `hitsumabushi_clock_linux.c`:
   - `hitsumabushi_nanotime1()`
   - `hitsumabushi_usleep()`
   - `hitsumabushi_walltime1()`
 - move the following functions to `hitsumabushi_cpu_linux.c`:
   - `hitsumabushi_sched_getaffinity()`
 - create a new `hitsumabushi_filesystem_linux.c` or `hitsumabushi_fs_linux.c` and move the following functions there:
   - `hitsumabushi_open()`
   - `hitsumabushi_read()`
   - `hitsumabushi_write1()`
   - `hitsumabushi_lseek()`
   - `hitsumabushi_fcntl()`
   - `hitsumabushi_fstat()`
   - `hitsumabushi_renameat()`
   - `hitsumabushi_fstatat()`
   - `hitsumabushi_closefd()`

This will only leave the following functions in `hitsumabushi_syscalls_linux.c`:
 - `hitsumabushi_exit()`
 - `hitsumabushi_gettid()`
 - `hitsumabushi_osyield()`

These functions are very threading- and process-related, maybe they could be merged together with the `futex` functions into something like `hitshumabushi_thread_linux.c`? Just in case we want to limit the number of extra C files.

Also, the `1` suffix in `c_write1()` is only here because there was already a `write()` function in the Go runtime. I would like to suggest dropping the suffix and go for `hitsumabushi_write()` (same for the other similar function), because it’s really just an implementation detail that the user doesn’t really need to know about.

**Comment by hajimehoshi** [comment]
> These functions are very threading- and process-related, maybe they could be merged together with the futex functions into something like hitshumabushi_thread_linux.c? Just in case we want to limit the number of extra C files.

Sounds a good idea! This would break compatibility but that's fine as this is still v0.

> Also, the 1 suffix in c_write1() is only here because there was already a write() function in the Go runtime. I would like to suggest dropping the suffix and go for hitsumabushi_write() (same for the other similar function), because it’s really just an implementation detail that the user doesn’t really need to know about.

write1 is a very special function to dump a message from panic. So I recommend keeping 1. In general, I prefer keeping the Go function names whenever possible.

https://cs.opensource.google/go/go/+/master:src/runtime/os_windows.go;l=706-711;drc=e252dcf9d38ce9192bccacb7e33867cbfbd22b6c

```
// write1 must be nosplit because it's used as a last resort in
// functions like badmorestackg0. In such cases, we'll always take the
// ASCII path.
```

**Comment by samhocevar** [comment]
I think the work is done! Now `overlay.go` has the following replacement functions:
 - `func ClockFilePath(os string) (string, error)`
 - `func ThreadsFilePath(os string) (string, error)`
 - `func FilesystemFilePath(os string) (string, error)`
 - `func MemoryFilePath(os string) (string, error)`
 - `func CPUFilePath(os string) (string, error)`

Maybe `ThreadsFilePath` and `CPUFilePath` could be merged even more? The rest looks quite well grouped by theme.

Also I noticed in https://github.com/hajimehoshi/hitsumabushi/pull/13/commits/1ca1ec100fcb38f4a0cf79a39a868b74d8d35585 that a lot of functions were actually not used, does that make sense to you?

File: 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c [hunk 5]
```diff
@@ -1,8 +1,23 @@
...
+  clock_gettime(CLOCK_REALTIME, &tp);
+  *sec = tp.tv_sec;
+  *nsec = tp.tv_nsec;
+}
+
+int64_t hitsumabushi_nanotime1() {
+  struct timespec tp;
+  clock_gettime(CLOCK_MONOTONIC, &tp);
+  return (int64_t)(tp.tv_sec) * 1000000000ll + (int64_t)tp.tv_nsec;
 }
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c:23]: This will require Nintendo Switch side, right? Did you need these functions ?

File: 1.19_linux/runtime/cgo/hitsumabushi_cpu_linux.c [hunk 6]
```diff
@@ -1,8 +1,20 @@
...
-#include <stdint.h>
+#include <stddef.h> // for size_t
+#include <stdint.h> // for int32_t
+#include <sched.h> // for pid_t
 
 int32_t hitsumabushi_getproccount() {
 	return 1;
 }
+
+int32_t hitsumabushi_sched_getaffinity(pid_t pid, size_t cpusetsize, void *mask) {
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/cgo/hitsumabushi_cpu_linux.c:12]: Do we really need to expose this for rewriting?

File: 1.19_linux/runtime/time_linux_amd64.s.patch [hunk 7]
```diff
@@ -0,0 +1,4 @@
+//--from
+//go:build !faketime
+//--to
+//go:build !faketime && false
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/time_linux_amd64.s.patch:4]: Just `//go:build ignore`?

**Review by hajimehoshi — COMMENTED** [review]
I'm a little skeptical about exposing some functions as commented. Did you actually need to overwrite these functions on your platform?

**Comment by hajimehoshi** [comment]
Yes, mmap and signals are not available on Nintendo switch.

2022年11月18日(金) 21:03 Sam Hocevar ***@***.***>:

File: 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c (hunk 5 above)
Comment (samhocevar) [reply 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c:23]: I am not sure I understand the question. The above works fine for me on Switch. Could that be because I use `Compilers/NX/nx/aarch64/bin/clang.exe` but you use a Linux compiler?

File: 1.19_linux/runtime/cgo/hitsumabushi_cpu_linux.c (hunk 6 above)
Comment (samhocevar) [reply 1.19_linux/runtime/cgo/hitsumabushi_cpu_linux.c:12]: Probably not required, indeed. The Sony consoles have an equivalent for this function so I was happy to use it, but it’s really not very important.

File: 1.19_linux/runtime/time_linux_amd64.s.patch (hunk 7 above)
Comment (samhocevar) [reply 1.19_linux/runtime/time_linux_amd64.s.patch:4]: OK!

**Comment by samhocevar** [comment]
> Yes, mmap and signals are not available on Nintendo switch.

I would say that they belong to a file outside of Hitsumabushi, then. I think you need them because your compiler generates dependencies to `mmap` etc., but when using the Nintendo or Switch compilers, these functions are not needed, and at least on the PlayStation they cause conflicts with the system’s libraries.

File: 1.19_linux/runtime/sys_linux_amd64.s.patch [hunk 8]
```diff
@@ -21,7 +21,7 @@ TEXT runtime·open_trampoline(SB),NOSPLIT,$0
 	MOVL	12(DI), DX		// arg 3 mode
 	MOVQ	0(DI), DI		// arg 1 pathname
 	XORL	AX, AX			// vararg: say "no float args"
-	CALL	c_open(SB)
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/sys_linux_amd64.s.patch:24]: I'm afraid this might be a burden on you, but I think it'd be nice to have a separate PR to rename the existing `c_` functions to `hitsumabushi_` functions

File: 1.19_linux/runtime/cgo/gcc_linux_arm64.c.patch [hunk 9]
```diff
@@ -28,23 +28,17 @@ void
...
-	pthread_sigmask(SIG_SETMASK, &ign, &oset);
-
 	pthread_attr_init(&attr);
 	pthread_attr_setstacksize(&attr, 16 * 4096); // Hack for some special environments
 	pthread_attr_getstacksize(&attr, &size);
 	// Leave stacklo=0 and set stackhi=size; mstart will do the rest.
 	ts->g->stackhi = size;
 	err = _cgo_try_pthread_create(&p, &attr, threadentry, ts);
 
-	pthread_sigmask(SIG_SETMASK, &oset, nil);
```
Comment (hajimehoshi) [inline 1.19_linux/runtime/cgo/gcc_linux_arm64.c.patch:46]: Removing pthread functions can be a pure refactoring in an independent PR.

File: 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c (hunk 5 above)
Comment (hajimehoshi) [reply 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c:23]: Let's talk in the Discord.

**Comment by samhocevar** [comment]
I think this PR can be abandoned. It has been split into many smaller PRs (#17, #19, #20) and the remaining differences are actually not very important because there is no use case for overriding all these other functions (at least for now).

**Comment by hajimehoshi** [comment]
Sure, let's revisit this when we find more functions we want to overwrite.
